### PR TITLE
Implement Hint#48 pack512

### PR DESCRIPTION
**Pull request by pefontana**
# Implement Hint#48 pack512

## Description
issue: #979 
#
Implement hint:
```
                "def pack_512(u, num_bits_shift: int) -> int:",
                "    limbs = (u.d0, u.d1, u.d2, u.d3)",
                "    return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))",
                "",
                "x = pack_512(ids.x, num_bits_shift = 128)",
                "p = ids.p.low + (ids.p.high << 128)",
                "x_inverse_mod_p = pow(x,-1, p)",
                "",
                "x_inverse_mod_p_split = (x_inverse_mod_p & ((1 << 128) - 1), x_inverse_mod_p >> 128)",
                "",
                "ids.x_inverse_mod_p.low = x_inverse_mod_p_split[0]",
                "ids.x_inverse_mod_p.high = x_inverse_mod_p_split[1]"
```

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [x] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [x] CHANGELOG has been updated.

### CODE REVIEW

**Comment by Oppen** [comment]
1000! 🎉 🎉 🎉

File: src/hint_processor/builtin_hint_processor/vrf/pack_512.rs [hunk 1]
```diff
@@ -19,13 +22,13 @@ def pack_512(d0, d1,d2,d3, num_bits_shift: int) -> int:
...
-    for (i, limb) in limbs.iter().enumerate() {
-        result += limb << (num_bits_shift * i);
-    }
-    result
+fn pack_512(limbs: &[Felt252; 4], num_bits_shift: usize) -> BigInt {
+    #[allow(deprecated)]
+    limbs
+        .iter()
+        .enumerate()
+        .map(|(idx, value)| value.to_bigint().shl(idx * num_bits_shift))
```
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/vrf/pack_512.rs:30]: Unintuitively, `Felt252::to_bigint` alters the value and is probably not what you want here. Use `Felt252::to_biguint`. If you need it to be signed you can use `BigUint::to_bigint` on the result.
The reason for this is that it turns them into the signed version of the field, that is, it does `P - x` when `x > P / 2`.

File: cairo_programs/garaba_programs/u255.cairo
[the remark this answers is not in the document]
Comment (pefontana) [reply cairo_programs/garaba_programs/u255.cairo:1]: yes, sorry the PR was in draft and I was making some fast implementation.
I end up with `cairo_programs/inv_mod_p_uint512.cairo` and `cairo_programs/fq_uint256.cairo`

File: cairo_programs/garaba_programs/u255.cairo [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+    let res_high = a.high + b.high;
+    %{
+        has_carry_low = 1 if ids.res_low >= ids.SHIFT else 0
+        ids.has_carry_low=has_carry_low
+        ids.has_carry_high = 1 if (ids.res_high+has_carry_low) >= ids.SHIFT else 0
+    %}
+
+    if (has_carry_low != 0) {
+        if (has_carry_high != 0) {
+            %{ print("case 0") %}
```
[the remark this answers is not in the document]
Comment (pefontana) [reply cairo_programs/garaba_programs/u255.cairo]: This is a library that I copy pasted from `cairo_programs/fq_uint256.cairo`.  the program we execute is `cairo_programs/inv_mod_p_uint512.cairo`

File: src/hint_processor/builtin_hint_processor/vrf/pack_512.rs (hunk 1 above)
Comment (pefontana) [reply src/hint_processor/builtin_hint_processor/vrf/pack_512.rs]: Thanks for this catch!

File: src/hint_processor/builtin_hint_processor/vrf/pack_512.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+pub fn inv_mod_p_uint512(
+    vm: &mut VirtualMachine,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let limbs_ptr = get_relocatable_from_var_name("x", vm, ids_data, ap_tracking)?;
+    let limbs: Vec<Felt252> = vm
+        .get_integer_range(limbs_ptr, 4)?
+        .iter()
+        .map(|f| f.clone().into_owned())
```
[the remark this answers is not in the document]
Comment (pefontana) [reply src/hint_processor/builtin_hint_processor/vrf/pack_512.rs]: mmm, I tried Cow::into_owned and didn't work.

**Review by Oppen — APPROVED** [review]

File: src/hint_processor/builtin_hint_processor/vrf/inv_mod_p_uint512.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    let x_inverse_mod_p =
+        Felt252::from(div_mod(&BigInt::one(), &BigInt::from(x), &BigInt::from(p)));
+
+    let x_inverse_mod_p_ptr =
+        get_relocatable_from_var_name("x_inverse_mod_p", vm, ids_data, ap_tracking)?;
+
+    vm.insert_value(
+        x_inverse_mod_p_ptr,
+        &x_inverse_mod_p
+            & &Felt252::from_str_radix("340282366920938463463374607431768211455", 10).unwrap(),
```
Comment (MegaRedHand) [inline src/hint_processor/builtin_hint_processor/vrf/inv_mod_p_uint512.rs:85]: To make it easier to understand where the number comes from, we could change it to:
```suggestion
            & &Felt252::from(u128::MAX),
```

**Review by MegaRedHand — APPROVED** [review]
LGTM

File: src/hint_processor/builtin_hint_processor/vrf/inv_mod_p_uint512.rs (hunk 4 above)
Comment (pefontana) [reply src/hint_processor/builtin_hint_processor/vrf/inv_mod_p_uint512.rs]: Thanks tommy!

**Review by Oppen — APPROVED** [review]

**Pull request merged**